docs/wallets.md

```diff
@@ -0,0 +1,57 @@
+* beyond correctness of the code, there is little risk (no need to source randomness or execute potentially variable-time algorithms).
+
+Supporting seed generation is a little more involved so the tradeoff between
+implementation complexity and user value is less clear, especially since the
+Codex provides users instructions on doing generation themselves.
+
+## Import Support
+
+codex32 shares may be any length between 128 and 512 bits.
+Wallets should support import of 128- or 256-bit seeds; other lengths are optional.
+
+128-bit seeds are 48 characters in length, including the `MS1` prefix.
+256-bit seeds are 74. For other bit-lengths, see the BIP.
+
+The process for entering shares is:
+
+1. The user should enter the first share. To the extent possible given screen limitations, when entering share data, data should be displayed in uppercase, with visually distinct four-character windows. The first four-character window should include the `MS1` prefix, which should be pre-filled.
+1. Once the first share is fully entered, the wallet should validate the checksum and header before accepting it.
+   * The user should not be able to enter mixed-case characters. The user must be able to enter all bech32 characters as well as `?` indicating an erasure. Wallets may allow users to enter non-bech32 characters, at their discretion. (This may be useful to guide error correction, e.g. by attempting to replace all `B`s with `8`s.)
+   * If the header is invalid, the wallet should highlight this and prevent the user from entering additional data until it is fixed. An invalid header is one that starts with a character other than `0` or `2` through `9`, or one which starts with `0` but whose share index is not `S`. For shares after the first, a header is also invalid if its threshold and identifier do not match those of the first share.
+   * If the checksum is invalid, the wallet may use an error-correction algorithm to determine a corrected share, but the wallet MUST show these corrections to the user rather than silently applying them.
+   * For substitution errors, the wallet may highlight the offending 4-character window or the offending character. It may also show the corrected character.
+   * If the wallet can determine insertion or deletion errors, it should highlight the offending 4-character window. When detecting insertion or deletion errors, the wallet may assume that the correct share length is a multiple of 16 bytes.
+1. If the share length is *not* 16, 32 or 64 bytes, but the checksum passes, the wallet should confirm that the user intends to import a non-standard share length.
+1. After the first share has been entered and accepted, the wallet now knows the seed ID and threshold value.
+   * If the first share had index `S`, this was the actual seed and the import process is complete.
+   * Otherwise, the first character of the share will be a numeric character between `2` and `9` inclusive. The user must enter this many shares in total.
```

Comment:
Do we need to support a case where the import progress can be asymmetrically encrypted, saved and resumed later by entering a "resume key"? For example Bails runs on a laptop, if the user is physically visiting share locations, the battery could die before a threshold is reached, not to mention issues of not powering the device off if crossing international borders.

Feel free to open a new issue for this, or I will if it's something wallets should support as it's an implementation detail of decent complexity, especially since flash memory cannot be relied upon to reliably erase itself due to wear leveling. (secure erase is done by physically shredding/burning the "resume key" afterwards)

Reply:
Hmm, I would like to support this. I did actually do such a process by hand and it was very useful to be able to do.

But I think of this document as a "minimum set of supported features", which is also why I make pretty weak recommendations for error correction etc. With the intention that wallets are free to try to add extra functionality. I'm not sure how best to express this.

Reply:
Mention the useful but optional features using the word "May" so developers can consider if they are worth implementing or not.

Error correction seems like a "should, whereever possible" feature since it makes a large UX improvement.

Electronics allow asymmetric encryption. The header of saved share progress can index the public key of the private 'resume key' for a particular set of shares being restored in multiple sessions. This avoids traveling to shares with the symmetric key that decrypts them all. This is a nice to have, but it's extra labor for users to write a restore key and not helpful for every recovery.

Reply:
Similar to how the BIP says wallets SHOULD display a corrected string. Without going into any details how to. One sentence: "Wallets MAY encrypt and store recovery progress, facilitating resumption of the process later by entering a designated 'resume key'."

I wouldn't even go into how to encrypt, seems device specific. I will eventually implement this in my project, I'll probably go the asymmetric encryption route. Then we can make a reference implementation for "save/resume"

This project uses SSS and had to solve similar problem. I gave them a few options and they went with: each time a share was gathered or transaction partially signed, they'd encrypt it by the pubkeys of all the remaining locations. So data in transit is not readable.
https://github.com/wild-kard/arctica/

I preferred writing a fresh resume key after the first save, keeping it somewhere safe, encrypting everything gathered by its public key, decrypting in a safe place to recover the secret and shredding it after.

Reply:
>Error correction seems like a "should, whereever possible" feature since it makes a large UX improvement.

Yeah, I think you're right. Though it is a big implementation burden which sucks. I guess having reference Python code will make the situation better.

> Wallets MAY encrypt and store recovery progress, facilitating resumption of the process later by entering a designated 'resume key'

I think even this much specification leads to complicated questions about key management. It's not obvious to me that the "asymmetric crypto with resume keys at each location" paradigm is the right one. I'm tempted to just say "Wallets MAY encrypt and store recovery progress, to allow recovery without having all shares available at once" and leaving it at that.

Reply:
I prefer a single resume key that's written down and kept somewhere safe before the user shuts down if they want to save progress. Then subsequent shares encrypt to its pubkey.

But if you think even mentioning it opens too much a can of worms, it's fine to omit, I'm sure almost every project of appreciable size will eventually have the open issue:  "I shut down while restoring and lost my progress, can we get a save feature"

Reply:
>But if you think even mentioning any implementation detail opens too much a can of worms, it's fine to omit, I'm sure almost every project of appreciable size will eventually have the open issue: "I shut down while restoring and lost my progress, can we get a save feature" and they can see what works for their hardware's threat model.

Yep. And I'm happy to revisit this doc later if some sort of consensus emerges about how best to do this.

>Without ECC, Codex32 could be harder to import than bip39, where at least mispelled semi-legible words have natural error correction. So I think it's important for it to be an all around superior replacement.

Definitely agreed. (Though BIP39 does have a few distance-one words so sometimes you don't get natural-word error correction. I have been burned by this with real funds before. Was **not** impressed after writing a bunch of python seed-grinding shit and loading real seed data into it, to find the spec had screwed me like this.)